spec/Classes.tex

```diff
@@ -663,73 +604,456 @@ \subsection{User-Defined Constructors}
+class LabeledPoint {
+  var x : real;
+  var y : real;
+  var msg : string = 'Unlabeled';
+
+  proc init(x: real, y: real) {
+    this.x = x;
+    this.y = y;
+    // compiler inserts "this.msg = 'Unlabeled'";
+  }
+
+  proc init(msg : string) {
+    // compiler inserts "this.x = 0.0;"
+    // compiler inserts "this.y = 0.0;"
+    this.msg = msg;
+  }
+}
+
+var A = new LabeledPoint(2.0, 3.0);
+var B = new LabeledPoint("Origin");
+\end{chapel}
+\begin{chapelpost}
+writeln(A);
+writeln(B);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 2.0, y = 3.0, msg = Unlabeled}
+{x = 0.0, y = 0.0, msg = Origin}
+\end{chapeloutput}
+The first initializer initializes the values of fields \chpl{x} and \chpl{y},
+and the compiler inserts initialization for the \chpl{msg} field by using
+its default value. The second initializer initializes the \chpl{msg} field,
+and the compiler inserts initialization for fields \chpl{x} and \chpl{y}
+based on the type of those fields (\rsec{Default_Values_For_Types}).
+\end{chapelexample}
+
+In order to reduce ambiguity and to ensure a well-defined order for
+side-effects, the language requires that all fields be initialized in field
+declaration order. This applies regardless of whether field initializations are
+omitted from the initializer body. If fields are initialized out of order, a
+compile-time error will be issued.
+
+\begin{chapelexample}{fieldsOutOfOrder.chpl}
+In the following code:
+\begin{chapel}
+class Point3D {
+  var x = 1.0;
+  var y = 1.0;
+  var z = 1.0;
+
+  proc init(x: real) {
+    this.x = x;
+    // compiler inserts "this.y = 1.0;"
+    this.z = y * 2.0;
+  }
+
+  proc init(x: real, y: real, z: real) {
+    this.x = x;
+    this.z = z;
+    this.y = y; // Error!
+  }
+}
+
+var A = new Point3D(1.0);
+var B = new Point3D(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+fieldsOutOfOrder.chpl:12: In initializer:
+fieldsOutOfOrder.chpl:17: error: Field "y" initialized out of order
+fieldsOutOfOrder.chpl:17: note: initialization of fields before .init() call must be in field declaration order
+\end{chapeloutput}
+
+The first initializer leverages the well-defined order of omitted field
+initialization to use the default value of field \chpl{y} in order to
+explicitly initialize field \chpl{z}.
+
+The second initializer initializes field \chpl{z} before field \chpl{y},
+causing a compile-time error to be issued.
+
+\end{chapelexample}
+
+\begin{rationale}
+Without this rule the compiler could insert default initialization for field
+\chpl{y} before \chpl{z} is explicitly initialized. The following statement
+would then be {\em assignment} to field \chpl{y}, despite appearing to be
+initialization. This subtle difference may be confusing and surprising, and
+is avoided by requiring fields to be initialized in field declaration order.
+\end{rationale}
+
+\subsubsection{The complete Method}
```

Comment:
There's a part of me that wonders if this section should be renamed.  Maybe something like "Limitations in Initializers and the complete Method"?  The section covers the need for `complete` but someone reading the initializers section of the spec might not realize they need to pay attention to it, especially if they are unfamiliar with the `complete` method in the first place

Reply:
"Calling Methods in Initializers" ? That doesn't really cover passing 'this' to functions though...

"Limitations on Instance Usage in Initializers" ? Feels wordy to me

Reply:
Going with "Limitations on Instance Usage in Initializers" for now.